topics/products/phpstorm/php_open_api_php_type_providers.md

```diff
@@ -0,0 +1,303 @@
+[//]: # (title: PHP Type Providers)
+
+<!-- Copyright 2000-2022 JetBrains s.r.o. and other contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file. -->
+
+Type inference in PhpStorm is built on top of type providers, each of which is responsible for inferring the types of
+specific PSI elements.
+For example, `com.jetbrains.php.lang.psi.resolve.types.PhpArrayAccessTP` is responsible for inferring the types of expressions like `$arr[10]`.
+There are dozens of such providers, and they all work one after another to provide type information when needed.
+
+All providers inherit from `com.jetbrains.php.lang.psi.resolve.types.PhpTypeProvider4`, which is registered in the `com.jetbrains.php.typeProvider4` extension point.
+
+## Types in PhpStorm
+
+The first stage of type inference in PhpStorm takes place at the indexing stage, at this moment PhpStorm don’t
+know anything about the functions that are in the code, don’t know what types they return, don’t know the classes
+and what methods are called, all this imposes restrictions of how types work in PhpStorm.
+
+Because of this, types in PhpStorm are divided into two types:
+
+- Complete types
+- Incomplete types
+
+Internally, types are stored in the index as strings, these strings can contain one type or several if it is a [Union Type](#union-types).
+
+### Complete types
+
+Complete types are types that are already known exactly, for example, in the following code:
```

Comment:
Complete types are types that are known exactly based on the content of the containing file